gluonnlp/data/sampler.py

```diff
@@ -63,6 +65,150 @@ def _bucket_average_lengths(bucket_sample_ids, seq_lengths):
+        r"""This generate bucket keys given that all the buckets have the same width.
+
+        Parameters
+        ----------
+        single_element : bool
+            Whether lengths of sequences is a list of int.
+        max_lengths : int of list of int
+            Maximum of lengths of sequences.
+        min_lengths : int of list of int
+            Minimum of lengths of sequences.
+        num_buckets : int
+            Number of buckets
+        """
+        if not single_element:
+            bucket_width_l = [max((max_len - min_len) // num_buckets, 1)
+                              for max_len, min_len in
+                              zip(max_lengths, min_lengths)]
+            bucket_keys = \
+                [tuple(max(max_len - i * width, min_len) for max_len, min_len, width in
+                       zip(max_lengths, min_lengths, bucket_width_l))
+                 for i in range(num_buckets)]
+        else:
+            bucket_width = max((max_lengths - min_lengths) // num_buckets, 1)
+            bucket_keys = [max(max_lengths - i * bucket_width, min_lengths)
+                           for i in range(num_buckets)]
+        return bucket_keys
+
+
+class LinearWidthBucket(BucketScheme):
+    r""" Buckets with linearly increasing width.
+    """
+    def __call__(self, single_element, max_lengths, min_lengths, num_buckets):
+        r"""This implements a scheme for linearly increasing bucket width:
+                :math:`w_i = \alpha * i + 1` for all :math:`i \geq 1`.
+
+        Parameters
+        ----------
+        single_element : bool
+            Whether lengths of sequences is a list of int.
+        max_lengths : int of list of int
+            Maximum of lengths of sequences.
+        min_lengths : int of list of int
+            Minimum of lengths of sequences.
+        num_buckets : int
+            Number of buckets
+        """
+        if not single_element:
+            alpha_l = [2 * float(max_len - min_len - num_buckets)
+                       / (num_buckets * (num_buckets + 1))
+                       for max_len, min_len in
+                       zip(max_lengths, min_lengths)]
+            bucket_keys = \
+                [tuple(int(round(min_len + alpha * (((i + 1) * (i + 2)) / 2) + i + 1))
+                       for min_len, alpha in zip(min_lengths, alpha_l))
+                 for i in range(num_buckets)]
+            bucket_keys[-1] = tuple(max(max_bucket_key, max_len)
+                                    for max_bucket_key, max_len
+                                    in zip(bucket_keys[-1], max_lengths))
+        else:
+            alpha = 2 * float(max_lengths - min_lengths - num_buckets) \
+                    / (num_buckets * (num_buckets + 1))
+            bucket_keys = [int(round(min_lengths + alpha * (((i + 1) * (i + 2)) / 2) + i + 1))
+                           for i in range(num_buckets)]
+            bucket_keys[-1] = max(bucket_keys[-1], max_lengths)
```

Comment:
Should we apply the filter `max(ele, max_lengths)` to all the bucket_keys?

Reply:
No, we only need to make sure the last one is at least equal to max_lengths. If it is applied to all the bucket keys, then every key will equal max_lengths.

Reply:
I see. I was considering the case when the bucket_keys could be larger than the max_lengths. So we  will waste some resources.

Reply:
Theoretically, the difference will be within 1.